extends Node
class_name PersitDataBridge

static var persistCalback: Dictionary[String, Callable] = {}

static var waitClientReady: Array[Callable] = []

static func setup_persistence_manager(calback: Callable):
	if calback.is_null(): 
		printerr("callable is null !!")
	if OS.has_feature("dedicated_server"):
		var pm = PersistanceManager
		if not pm:
			push_error("PersistanceManager is null!")
			return
		# Connect all PersistanceManagerSignal
		if not pm.SaveCompleted.is_connected(_on_save_completed):
			pm.SaveCompleted.connect(_on_save_completed)
		
		if not pm.DeleteCompleted.is_connected(_on_delete_completed):
			pm.DeleteCompleted.connect(_on_delete_completed)
		
		if not pm.QueryCompleted.is_connected(_on_query_completed):
			pm.QueryCompleted.connect(_on_query_completed)
		
		if not pm.FindByIdCompleted.is_connected(_on_find_by_id_completed):
			pm.FindByIdCompleted.connect(_on_find_by_id_completed)
		
		# DÃ©marrer les opÃ©rations si le manager est prÃªt
		if pm.IsReady:
			print("âœ… Client Alredy start")
			calback.call()
		else:
			print("â³ Waiting signal ClientReady...")
			if not pm.ClientReady.is_connected(_on_client_ready):
				pm.ClientReady.connect(_on_client_ready)
			waitClientReady.push_back(calback)
	else :
		print("only dev for the server")

# ============ EVENT HANDLERS ============
static func _on_client_ready():
	print("ğŸš€ Signal ClientReady !")
	while waitClientReady.size() > 0:
		var calback = waitClientReady.pop_back()
		if !calback.is_null():
			calback.call()

static func _on_save_completed(success: bool, uid: String, error_message: String, request_id: String):
	print("ğŸ’¾ Save completed - RequestID: ", request_id)
	if success:
		if uid == "":
			print("âœ…Object Updatated ")
		else :
			print("âœ…Object save with UID: ", uid)
		if persistCalback.has(request_id):
			persistCalback[request_id].call(uid)
			persistCalback.erase(request_id)
		else : 
			printerr("no callback for this request id ")
	else:
		printerr("âŒ Failed save: ", error_message)

static func _on_delete_completed(success: bool, error_message: String, request_id: String):
	print("ğŸ—‘ï¸ Delete completed - RequestID: ", request_id)
	if success:
		print("âœ… Objet deleted succÃ¨s")
		if persistCalback.has(request_id):
			persistCalback[request_id].call()
			persistCalback.erase(request_id)
		else : 
			printerr("no callback for this request id ")
	else:
		printerr("âŒ Failed delete: ", error_message)

static func _on_query_completed(success: bool, json_data: String, error_message: String, request_id: String):
	print("ğŸ” Query completed - RequestID: ", request_id)
	if success:
		print("âœ… RequÃªte Success")
		if persistCalback.has(request_id):
			persistCalback[request_id].call(json_data)
			persistCalback.erase(request_id)
		else : 
			printerr("no callback for this request id ")
	else:
		printerr("âŒ Ã‰chec requÃªte: ", error_message)

static func _on_find_by_id_completed(success: bool, json_data: String, error_message: String, request_id: String):
	print("ğŸ¯ FindById completed - RequestID: ", request_id)
	if success:
		print("âœ… Serach By ID success")
		if persistCalback.has(request_id):
			persistCalback[request_id].call(json_data)
			persistCalback.erase(request_id)
		else : 
			printerr("no callback for this request id ")
	else:
		printerr("âŒ Ã‰chec recherche: ", error_message)



# ============ function for external use ============
static func save_data(data: DataObject,  calback: Callable):
	var pm = PersistanceManager
	if pm and pm.IsReady:
		var rid = pm.StartSaveAsync(data.serialize())
		persistCalback[rid]=calback
	else:
		printerr("âŒ PersistanceManager is not ready for save_data")


static func background_save_data(data: DataObject,  priority: int):
	var pm = PersistanceManager
	if pm and pm.IsReady:
		pm.BackgroundSaveObjAsync(data.serialize(),priority)
	else:
		printerr("âŒ PersistanceManager is not ready for save_data")

static func delete_data(uid: String, calback: Callable):
	if uid == "":
		print("âŒ not uid")
		return
	
	var pm = PersistanceManager
	if pm and pm.IsReady:
		var rid = pm.StartDeleteAsync(uid)
		persistCalback[rid]=calback
	else:
		printerr("âŒ PersistanceManager is not ready pour delete_data")

static func background_delete_data(uid: String,  priority: int):
	if uid == "":
		print("âŒ not uid")
		return
	var pm = PersistanceManager
	if pm and pm.IsReady:
		pm.BackgroundDeleteObjAsync(uid,priority)
	else:
		printerr("âŒ PersistanceManager is not ready for delete_data")
		
static func find_data_by_id(uid: String,calback: Callable):
	var pm = PersistanceManager
	if pm and pm.IsReady:
		var rid = pm.StartFindByIdAsync(uid)
		persistCalback[rid]=calback
	else:
		printerr("âŒ PersistanceManager is not ready find_data_by_id")

static func execute_custom_query(query_string: String,calback: Callable):
	var pm = PersistanceManager
	if pm and pm.IsReady:
		var rid = pm.StartQueryAsync(query_string)
		persistCalback[rid]=calback
	else:
		printerr("âŒ PersistanceManager is not ready execute_custom_query")
